### PR TITLE
[LLD][AArch64][ARM] Delay adding long thunk mapping symbols

### PR DESCRIPTION
**Pull request by smithp35**
When we create a thunk we don't know whether it will be short or long. Move the emission of the long thunk mapping symbol to when we transition to a long thunk. This improves disassembly and binary analysis as tools like BOLT identify thunks by disassembly.

### CODE REVIEW

**Comment by smithp35** [comment]
This removes the FIXME in the newly added aarch64-thunk-bti-multipass.s test. I've updated all the Arm thunks that have additional mapping symbols using the same pattern. I've not written any additional tests for all the Arm as these are quite difficult to write. I can add tests for some or all of these if necessary.

File: lld/ELF/Thunks.cpp [hunk 1]
```diff
@@ -69,6 +69,7 @@ class AArch64Thunk : public Thunk {
 private:
   bool mayUseShortThunk = true;
   virtual void writeLong(uint8_t *buf) = 0;
+  virtual void addLongMapSyms() {}
```
Comment (MaskRay) [inline lld/ELF/Thunks.cpp:72]: Perhaps add some comment, something like: mapping symbols with non-zero offsets may become needed when mayUseShortThunk becomes false later.

**Comment by MaskRay** [comment]
> This removes the FIXME in the newly added aarch64-thunk-bti-multipass.s test. I've updated all the Arm thunks that have additional mapping symbols using the same pattern. I've not written any additional tests for all the Arm as these are quite difficult to write. I can add tests for some or all of these if necessary.
> 
> I considered an alternative approach of adding a finalize() member function to ThunkSection and having that be called once per ThunkSection, that would then know for sure whether a thunk is short or long and could add the additional symbols. That would require another pass through all the input sections to find the ThunkSections though so unless we need finalize for something else I went with adding the additional symbols when we transition from short to long.

Thanks. The current approach looks good to me.

It may be worth mentioning #108989 and aarch64-thunk-bti-multipass.s in the description.

**Review by MaskRay — APPROVED** [review]

File: lld/ELF/Thunks.cpp (hunk 1 above)
Comment (smithp35) [reply lld/ELF/Thunks.cpp:76]: Thanks for the review, I've updated the comment. I'll also update the description as suggested below when merging.

**Pull request merged**